sourcegraph/sourcegraph
# Reset Cody Gateway usage cache automatically during limits sync on interval & TTL mismatch

- Closes https://github.com/sourcegraph/sourcegraph/issues/59418
- Closes https://github.com/sourcegraph/cody/issues/2475

## tldr;

1. The value of Cody usage counter is not reset to 0 when the user changes their plan. 
2. The TTL of Cody usage counter if not updated from 30 days to 1 day when the user upgrades from Free to Pro, or the other way around. 

## Slack
https://sourcegraph.slack.com/archives/C04MSD3DP5L/p1704375768295309?thread_ts=1704370864.536539&cid=C04MSD3DP5L

## Background
Cody Gateway maintains 2 caches:
1. limits config with
    -  24hrs TTL (controlled by `CODY_GATEWAY_SOURCES_CACHE_TTL`). 
    - But is is considered stale in the code if it is older than 15 minutes and thus we re-fetch every 15 mins. 
2. usage counter with 
    - 30 days TTL for free users (monthly rate limits)
    - [SHOULD BE, BUT NEVER UPDATED] 1-day TTL for pro users (daily rate limits)

Right now, when the user upgrades, we request Cody Gateway to refresh the limits config in its cache. But nothing is implemented to reset the usage counter or change its TLL, which is the BUG. 

## Solution

When Cody Gateway fetches limit config from dotcom source, whether on
1. cache miss
2. stale cache older than 15 mins
3. trigger by refresh config API request

it checks if  `newConfig.Interval != oldConfig.interval` and deletes the usage counter cache.

We will also update `CODY_GATEWAY_SOURCES_CACHE_TTL` in the config from 24hours to 31 or 62 days so that the limits config is persisted in the cache for longer and covers the complete billing cycle. 

## Other Improvements

### FIRST
There is also a logic in the Limiter which checks the following:
`if usage_cache.TTL > rate_limit.Interval` then update the cache's TTL = interval. 

The above logic is only a safeguard from corrupt usage cache or handle the edge cases during any kind of code change and migrations. 

This logic only worked on successful requests and not when the user is already hitting the rate limits. Which defeats the purpose of having it in the first place. 

As part of this PR, I have moved this check, so it is performed even when the request it already hitting rate limits.

### SECOND (TODO, will be done as part of a separate PR)
We will block users to downgrade on the same day they upgrade to prevent abuse until the payments are integrated. 

## Test plan
1. Follow this guide to setup Gateway Locally https://docs.sourcegraph.com/dev/how-to/cody_gateway
2. Additionally, add `CODY_GATEWAY_SOURCES_CACHE_TTL: 720h` (30 days) to `sg.config.overwrite.yaml` under `commands.cody-gateway.env`. This will store the actor config in the cache for 30 days. CURRENTLY 24h.
3. Go to `cmd/cody-gateway/internal/actor/dotcomuser/dotcomuser.go:40` and edit `defaultRefreshInterval` = 0. 
4. Run `sg start dotcom`
5. Make sure `cody-pro` feature flag is turned on
6. Connect a redis client to localhost:6379
7. Visit `/cody/manage` and verify that you are on Cody Free plan, if not, downgrade.
8. Visit Cody Web at `/cody/chat` and make a chat requests
9. Check for `v2:dotcom-users:v2:<user_id>` key in Redis: it stores the limits config in JSON.
      - The key should have a TTL in the neighbourhood of 25,92,000 seconds. (30 days as we updated the `CODY_GATEWAY_SOURCES_CACHE_TTL`).
      - Both `rateLimits.chat_completions.interval` and `rateLimits.code_completions.interval` should be exactly `2592000000000000` (30 days as the user is on the Free tier).
10. Check for`rate_limit:chat_completions:<user_id>` key in Redis: it stores the LLM requests usage counter.
      - The key should have a TTL in the neighbourhood of 25,92,000 seconds. (30 days as the user is on the Free tier).
      - The value should increment with each new chat message. 
11. Visit `/cody/manage` and upgrade to Cody Pro.
12.  Check for `v2:dotcom-users:v2:<user_id>` key in Redis:
      - Both `rateLimits.chat_completions.interval` and `rateLimits.code_completions.interval` should now be `86400000000000` (1 day as the user is on the Pro tier).
      - The TTL should still be in the neighourhood of 30 days. As controlled by the config.
13.  Check for`rate_limit:chat_completions:<user_id>` key in Redis.
      - The key be DELETED.
14. Visit `/cody/chat` and make a request.
15. Check for`rate_limit:chat_completions:<user_id>` key in Redis.
      - The key should have a TTL in the neighbourhood of 86,400 seconds. (1 day as the user is on the Pro tier).
      - The value should reset and will be 1 or 2 based on the number of chats you made after upgrading. 
16. Repeat the steps from 11-14 after downgrading back to Free. Accordingly check for the TLL and config numbers.
